### PR TITLE
Add LIMIT/OFFSET Support

### LINKED ISSUE #394: Add "Limit / Offset" support
Citus supports queries that involve ``LIMIT`` clauses, but not yet ``LIMIT / OFFSET``.

For these queries, we currently offer one of two workarounds:
1. Use cursors on the master node, or
2. Run ``CREATE TABLE AS (SELECT ... LIMIT 100)`` first, and then run ``OFFSET`` on that temporary table.

This issue tracks adding native ``LIMIT / OFFSET`` support to Citus.

### PR DESCRIPTION
**Pull request by aamederen**
Fixes #394 

This change adds LIMIT/OFFSET support for non router-plannable
distributed queries.

In cases that we can push the LIMIT down, we add the OFFSET value to
that LIMIT in the worker queries. When a query with LIMIT x OFFSET y is issued, 
the query is propagated to the workers as LIMIT (x+y) OFFSET 0, and on the 
master table, the original LIMIT and OFFSET values are used. With this change, 
we can use OFFSET wherever we can use LIMIT.

### CODE REVIEW

**Comment by ozgune** [comment]
@lfittl -- do you know why we need a signed CLA for this one?

**Comment by sumedhpathak** [comment]
@metdos Could you review this PR? Thanks!

**Comment by lfittl** [comment]
@ozgune We don't, but our team structure is odd and this confuses the CLA mechanism.

Its not a high priority to fix (since I've whitelisted the team members it affects), and we should probably fix our team setup instead of hacking the CLA tool.

File: src/test/regress/expected/multi_complex_expressions.out [hunk 1]
```diff
@@ -347,3 +347,84 @@ SELECT count(*) FROM lineitem JOIN orders ON l_orderkey = o_orderkey
...
+         97
+         98
+         99
+        100
+        101
+        102
+(10 rows)
+
+-- LIMIT/OFFSET with Subqueries
+SELECT
```
Comment (metdos) [inline src/test/regress/expected/multi_complex_expressions.out]: Do we support subqueries, or does this work because the subquery is flattened out?

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 2]
```diff
@@ -2220,6 +2222,24 @@ MakeIntegerConst(int32 integerValue)
 }
 
 
+/* Makes an 64-bit integer constant node from the given value, and returns that node. */
```
Comment (metdos) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: an -> a

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 3]
```diff
@@ -2220,6 +2222,24 @@ MakeIntegerConst(int32 integerValue)
...
+static Const *
+MakeIntegerConstInt64(int64 integerValue)
+{
+	const int typeCollationId = get_typcollation(INT4OID);
```
Comment (metdos) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: INT4OID -> INT8OID

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 4]
```diff
@@ -2220,6 +2222,24 @@ MakeIntegerConst(int32 integerValue)
...
+	const int16 typeLength = get_typlen(INT4OID);
+	const int32 typeModifier = -1;
+	const bool typeIsNull = false;
+	const bool typePassByValue = true;
+
+	Datum integerDatum = Int64GetDatum(integerValue);
```
Comment (metdos) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: maybe?
integerDatum -> integer64Datum
integerConst -> integer64Const

File: src/test/regress/expected/multi_complex_expressions.out [hunk 5]
```diff
@@ -347,3 +347,84 @@ SELECT count(*) FROM lineitem JOIN orders ON l_orderkey = o_orderkey
 SELECT count(*) FROM lineitem, orders WHERE l_orderkey + 1 = o_orderkey;
 ERROR:  cannot perform local joins that involve expressions
 DETAIL:  local joins can be performed between columns only
+-- Check that we can issue limit/offset queries
+-- SIMPLE OFFSET without LIMIT and ORDER BY
+SELECT o_orderkey FROM orders WHERE o_orderkey > 9000 AND o_orderkey < 9100 OFFSET 10;
```
Comment (metdos) [inline src/test/regress/expected/multi_complex_expressions.out]: Are we sure that this will give the same result all the time without an order by?

File: src/test/regress/sql/multi_complex_expressions.sql [hunk 6]
```diff
@@ -158,3 +158,33 @@ SELECT count(*) FROM lineitem JOIN orders ON l_orderkey = o_orderkey
...
+	li.l_partkey,
+	o.o_custkey,
+	li.l_quantity
+FROM
+	lineitem li JOIN orders o ON li.l_orderkey = o.o_orderkey 
+WHERE 
+	li.l_quantity > 25
+ORDER BY
+	li.l_quantity
+LIMIT 10 OFFSET 20;
```
Comment (metdos) [inline src/test/regress/sql/multi_complex_expressions.sql]: Missing new line at the end of the file.

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 7]
```diff
@@ -4135,6 +4160,26 @@ WorkerLimitCount(MultiExtendedOp *originalOpNode)
...
+		{
+			workerLimitCount = (Node *) MakeIntegerConstInt64(0);
+		}
+
+		workerLimitConst = (Const *) workerLimitCount;
+		workerNewLimitValueInt64 = DatumGetInt64(workerLimitConst->constvalue);
+		workerNewLimitValueInt64 += workerOffsetValueInt64;
+		workerLimitConst->constvalue = Int64GetDatum(workerNewLimitValueInt64);
+	}
+
```
Comment (metdos) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: Would something like this be simpler? We always sum limit and offset, and if the limit is null, use the initial value 0. 


```sql
	if (originalOpNode->limitOffset != NULL)
	{
		Const *workerOffsetConst = (Const *) originalOpNode->limitOffset;
		int64 workerOffsetValueInt64 = DatumGetInt64(workerOffsetConst->constvalue);
		int64 workerLimitValueInt64 = 0;

		/* if a limit is present, use it for initial value */
		if (workerLimitCount != NULL)
		{
			Const *workerLimitConst = (Const *) workerLimitCount;
			workerLimitValueInt64 = DatumGetInt64(workerLimitConst->constvalue);
		}

		workerLimitValueInt64 = workerLimitValueInt64 + workerOffsetValueInt64;
		workerLimitCount = (Node *) MakeIntegerConstInt64(workerLimitValueInt64);
	}
```

File: src/test/regress/expected/multi_complex_expressions.out (hunk 1 above)
Comment (aamederen) [inline src/test/regress/expected/multi_complex_expressions.out]: I have replaced the query with a more complex one. I think it still works.

File: src/backend/distributed/planner/multi_logical_optimizer.c (hunk 2 above)
Comment (aamederen) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: Changed.

File: src/backend/distributed/planner/multi_logical_optimizer.c (hunk 3 above)
Comment (aamederen) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: Changed.

File: src/backend/distributed/planner/multi_logical_optimizer.c (hunk 4 above)
Comment (aamederen) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: Changed.

File: src/test/regress/expected/multi_complex_expressions.out (hunk 5 above)
Comment (aamederen) [inline src/test/regress/expected/multi_complex_expressions.out]: No there is no such a guarentee. I have removed it from the tests.

File: src/test/regress/sql/multi_complex_expressions.sql (hunk 6 above)
Comment (aamederen) [inline src/test/regress/sql/multi_complex_expressions.sql]: Added.

File: src/backend/distributed/planner/multi_logical_optimizer.c (hunk 7 above)
Comment (aamederen) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: You are correct. I have also changed the code above this part for further simplification.

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 8]
```diff
@@ -4090,9 +4115,12 @@ WorkerLimitCount(MultiExtendedOp *originalOpNode)
 	List *groupClauseList = originalOpNode->groupClauseList;
 	List *sortClauseList = originalOpNode->sortClauseList;
 	List *targetList = originalOpNode->targetList;
+	bool orderByNonAggregates = !(HasOrderByAggregate(sortClauseList, targetList));
```
Comment (metdos) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: HasOrderByAggregate() checks if we have an order by an aggregate function. The opposite is not necessarily an order by without aggregates. There could be no order by at all. 

I checked other usages and the place you copy this. It always sits inside a if (sortClauseList != NIL) block, so we know that there is an order by clause at least.

Comment (metdos) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: You have this line below
(sortClauseList != NIL && orderByNonAggregates)

but still orderByNonAggregates has a wrong meaning in this case.

**Comment by metdos** [comment]
@aamederen maybe we can add regression tests with explain for the fix of #606 as we talked. I will do the final review after I come back from the vacation :)

**Comment by aamederen** [comment]
@metdos I have added the EXPLAIN statements showing the LIMITs pushed to the shards. Since EXPLAIN outputs are JOBID sensitive, I had to extract multi_complex_expressions from its parallel group.

File: src/test/regress/sql/multi_complex_expressions.sql [hunk 9]
```diff
@@ -158,3 +158,46 @@ SELECT count(*) FROM lineitem JOIN orders ON l_orderkey = o_orderkey
 -- Check that we make sure local joins are between columns only.
 
 SELECT count(*) FROM lineitem, orders WHERE l_orderkey + 1 = o_orderkey;
+
+-- Check that we can issue limit/offset queries
+
+-- SIMPLE LIMIT/OFFSET with ORDER BY
```
Comment (metdos) [inline src/test/regress/sql/multi_complex_expressions.sql]: SIMPLE -> Simple

File: src/test/regress/sql/multi_complex_expressions.sql [hunk 10]
```diff
@@ -158,3 +158,46 @@ SELECT count(*) FROM lineitem JOIN orders ON l_orderkey = o_orderkey
...
+SELECT o_orderkey FROM orders ORDER BY o_orderkey LIMIT 10 OFFSET 20;
+
+-- LIMIT/OFFSET with a Subquery
```
Comment (metdos) [inline src/test/regress/sql/multi_complex_expressions.sql]: Subquery -> subquery

File: src/test/regress/sql/multi_complex_expressions.sql [hunk 11]
```diff
@@ -158,3 +158,46 @@ SELECT count(*) FROM lineitem JOIN orders ON l_orderkey = o_orderkey
...
+SET citus.task_executor_type TO 'task-tracker';
+SELECT 
+	t.o_custkey,
+	COUNT(*) AS ss
+FROM 
+	(SELECT o_custkey, COUNT(*) AS order_count FROM orders GROUP BY o_custkey) t
```
Comment (metdos) [inline src/test/regress/sql/multi_complex_expressions.sql]: Can we use a more descriptive name for subquery instead of just "t"?
subtable, customer_count etc.

File: src/test/regress/sql/multi_complex_expressions.sql [hunk 12]
```diff
@@ -158,3 +158,46 @@ SELECT count(*) FROM lineitem JOIN orders ON l_orderkey = o_orderkey
...
+
+SET citus.task_executor_type TO 'real-time';
+
+-- Ensure that we push down LIMIT and OFFSET properly
+
+EXPLAIN (COSTS FALSE) SELECT o_custkey FROM orders LIMIT 10 OFFSET 15;
+
+EXPLAIN (COSTS FALSE) SELECT o_custkey FROM orders GROUP BY o_custkey LIMIT 10 OFFSET 15;
+
+EXPLAIN (COSTS FALSE) SELECT o_custkey FROM orders GROUP BY o_custkey ORDER BY o_custkey LIMIT 10 OFFSET 15;
```
Comment (metdos) [inline src/test/regress/sql/multi_complex_expressions.sql]: These two lines are longer than 90 characters.

File: src/test/regress/expected/multi_complex_expressions.out [hunk 13]
```diff
@@ -347,3 +347,155 @@ SELECT count(*) FROM lineitem JOIN orders ON l_orderkey = o_orderkey
...
+        100
+        101
+        102
+(10 rows)
+
+-- LIMIT/OFFSET with a Subquery
+SET citus.task_executor_type TO 'task-tracker';
+SELECT 
+	t.o_custkey,
+	COUNT(*) AS ss
```
Comment (metdos) [inline src/test/regress/expected/multi_complex_expressions.out]: Maybe, for more colorful results, group on another column in subquery and get the sum in the outer query :)

File: src/test/regress/expected/multi_complex_expressions.out [hunk 14]
```diff
@@ -347,3 +347,155 @@ SELECT count(*) FROM lineitem JOIN orders ON l_orderkey = o_orderkey
...
+   ->  Task
+         Node: host=localhost port=57637 dbname=regression
+         ->  Limit
+               ->  Seq Scan on orders_290006 orders
+ Master Query
+   ->  Limit
+         ->  Seq Scan on pg_merge_job_420038
+(11 rows)
+
+EXPLAIN (COSTS FALSE) SELECT o_custkey FROM orders GROUP BY o_custkey LIMIT 10 OFFSET 15;
```
Comment (metdos) [inline src/test/regress/expected/multi_complex_expressions.out]: Ohh, with the introduction of COSTS FALSE, we can't see the row numbers pulled from worker nodes. We can remove these tests.

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 15]
```diff
@@ -4080,19 +4100,27 @@ EqualOpExpressionLists(List *firstOpExpressionList, List *secondOpExpressionList
...
 static Node *
 WorkerLimitCount(MultiExtendedOp *originalOpNode)
 {
-	Node *workerLimitCount = NULL;
+	Node *workerLimitNode = NULL;
 	List *groupClauseList = originalOpNode->groupClauseList;
 	List *sortClauseList = originalOpNode->sortClauseList;
 	List *targetList = originalOpNode->targetList;
+	bool hasOrderByAggregate = HasOrderByAggregate(sortClauseList, targetList);
+	bool canPushDownLimit = false;
```
Comment (metdos) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: very very minor! Can we replace the order of canPushDownLimit and canApproximate? I think it will look better :)

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 16]
```diff
@@ -4102,40 +4130,56 @@ WorkerLimitCount(MultiExtendedOp *originalOpNode)
 	 * aggregates, we can push down the original limit. Else if we have order by
 	 * clauses with commutative aggregates, we can push down approximate limits.
 	 */
-	if (groupClauseList == NIL)
+	if (groupClauseList == NIL || sortClauseList == NIL || !hasOrderByAggregate)
```
Comment (metdos) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: I don't know if we changed the code here or I missed it before, but anyway, I think we have a problem here.

If we run this query on the master, it will push down limit to worker nodes, and this would cause wrong results.

I think that previously we didn't push down limit in this case.

select l_orderkey, count(*) from lineitem group by l_orderkey limit 10;

A set of results would be;
postgres=# select l_orderkey, count(*) from lineitem group by l_orderkey limit 10;
 l_orderkey | count
------------+-------
        129 |     7
       2848 |     5
       2054 |     7
       3623 |     7
        195 |     4
       1509 |     7
        705 |     2
       2918 |     1
        480 |     1
       1126 |     3
(10 rows)


But there would some "129" on different worker nodes which are left outside due to pushed down limit.

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 17]
```diff
@@ -4102,40 +4130,56 @@ WorkerLimitCount(MultiExtendedOp *originalOpNode)
...
+
+	/*
+	 * If offset clause is present and limit can be pushed down (whether exactly or
+	 * approximately), add the offset value to limit on workers
+	 */
+	if (originalOpNode->limitOffset != NULL && (canPushDownLimit || canApproximate))
+	{
+		Const *workerOffsetConst = (Const *) originalOpNode->limitOffset;
+		int64 workerOffsetValue = DatumGetInt64(workerOffsetConst->constvalue);
+		int64 workerLimitValue = 0;
```
Comment (metdos) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: Maybe workerOffsetCount and workerLimitCount for consistency in the function.

File: src/backend/distributed/planner/multi_logical_optimizer.c (hunk 16 above)
Comment (aamederen) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: We have discussed this before. The problem here is that, we do not specify any order and there is no guarantee on the result set if no order-by is specified.

Comment (metdos) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: > We have discussed this before. The problem here is that, we do not specify any order and there is no guarantee on the result set if no order-by is specified.

I think we discussed this if there is not any aggregate in the query like you noted in #606, then it makes sense. But, if there is an aggregate, we can return the wrong aggregate value. Does this make sense?

File: src/backend/distributed/planner/multi_logical_optimizer.c (hunk 17 above)
Comment (aamederen) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: Changed accordingly

File: src/backend/distributed/planner/multi_logical_optimizer.c (hunk 16 above)
Comment (aamederen) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: Yeah, makes sense. I have added a check for queries with aggregates but no `ORDER BY`. Do you have any other conditions that we should not push `LIMIT` down?

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 18]
```diff
@@ -4080,62 +4101,128 @@ EqualOpExpressionLists(List *firstOpExpressionList, List *secondOpExpressionList
  * WorkerLimitCount checks if the given extended node contains a limit node, and
  * if that node can be pushed down. For this, the function checks if this limit
  * count or a meaningful approximation of it can be pushed down to worker nodes.
- * If they can, the function returns the limit count. Otherwise, the function
+ * If they can, the function returns the limit count.
+ *
+ * The limit push-down decision tree is as follows:
+ *                                 group by?
```
Comment (metdos) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: Nice sketch 👍

**Comment by metdos** [comment]
Could you run citus_indent?

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 19]
```diff
@@ -4080,62 +4101,128 @@ EqualOpExpressionLists(List *firstOpExpressionList, List *secondOpExpressionList
...
-	Node *workerLimitCount = NULL;
+	Node *workerLimitNode = NULL;
 	List *groupClauseList = originalOpNode->groupClauseList;
 	List *sortClauseList = originalOpNode->sortClauseList;
 	List *targetList = originalOpNode->targetList;
+	bool hasOrderByAggregate = HasOrderByAggregate(sortClauseList, targetList);
+	bool canApproximate = CanPushDownLimitApproximate(sortClauseList, targetList);
+	bool hasAggregateTarget = HasAggregateTarget(targetList);
+	bool canPushDownLimit = false;
+	bool hasAggregateWithNoSort = groupClauseList != NIL && hasAggregateTarget &&
```
Comment (metdos) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: For boolean values, we prefer if blocks to be more explicit.

bool hasAggregateWithNoSort = false;

if (groupClauseList != NIL && hasAggregateTarget && 
 								  sortClauseList == NIL)
{
 hasAggregateWithNoSort = true;
}

Comment (metdos) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: Also naming of hasAggregateWithNoSort only implies (hasAggregateTarget && sortClauseList == NIL).

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 20]
```diff
@@ -4080,62 +4101,128 @@ EqualOpExpressionLists(List *firstOpExpressionList, List *secondOpExpressionList
...
+ * If they can, the function returns the limit count.
+ *
+ * The limit push-down decision tree is as follows:
+ *                                 group by?
+ *                              1/           \0
+ *                          order by?         (exact pd)
+ *                       1/           \0
+ *           has order by agg?           has aggregate
+ *            1/           \0            1/           \0
+ *     can approximage?    (exact pd)   (no pd)     (exact pd)
```
Comment (metdos) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: approximage -> approximate

File: src/backend/distributed/planner/multi_logical_optimizer.c (hunk 19 above)
Comment (metdos) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: This becomes a bit hard to follow. Instead can we create an if else block which represents your sketch above? We can leave "can approximate?" separate to keep 3 levels of indentation, but if it becomes hard to follow, we can even make an exception here.

if (group by)
{
    if (order by?)
    {
    }
  .... 
}
else
{
   canPushDownLimit = true;
}

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 21]
```diff
@@ -4080,19 +4100,39 @@ EqualOpExpressionLists(List *firstOpExpressionList, List *secondOpExpressionList
...
 	List *sortClauseList = originalOpNode->sortClauseList;
 	List *targetList = originalOpNode->targetList;
+	bool hasOrderByAggregate = HasOrderByAggregate(sortClauseList, targetList);
+	bool hasAggregateTarget = contain_agg_clause((Node *) targetList);
+	bool canPushDownLimit = false;
+	bool canApproximate = false;
 
 	/* no limit node to push down */
-	if (originalOpNode->limitCount == NULL)
+	if (originalOpNode->limitCount == NULL && originalOpNode->limitOffset == NULL)
```
Comment (metdos) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: I just noticed that if limit is null, we should not send any limit to worker nodes even though there is an offset value. Does this make sense?

This will also change some code path below around where we add offset to limit.

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 22]
```diff
@@ -4104,38 +4144,70 @@ WorkerLimitCount(MultiExtendedOp *originalOpNode)
...
+		canPushDownLimit = true;
 	}
-	else if (sortClauseList != NIL)
+	else if (sortClauseList == NIL)
 	{
-		bool orderByNonAggregates = !(HasOrderByAggregate(sortClauseList, targetList));
-		bool canApproximate = CanPushDownLimitApproximate(sortClauseList, targetList);
-
-		if (orderByNonAggregates)
+		if (!hasAggregateTarget)
```
Comment (metdos) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: This is the only place which we use hasAggregateTarget. We can move declaration to here.

File: src/test/regress/sql/multi_complex_expressions.sql [hunk 23]
```diff
@@ -158,3 +158,64 @@ SELECT count(*) FROM lineitem JOIN orders ON l_orderkey = o_orderkey
...
+-- Order by aggregate -> No push-down
+CREATE TEMP TABLE temp_limit_test_4 AS
+SELECT o_custkey, COUNT(*) AS ccnt FROM orders GROUP BY o_custkey ORDER BY ccnt DESC LIMIT 10 OFFSET 15;
+
+-- Has aggregate but no order-by -> No push-down
+CREATE TEMP TABLE temp_limit_test_5 AS
+SELECT o_custkey, COUNT(*) AS ccnt FROM orders GROUP BY o_custkey LIMIT 10 OFFSET 15;
+
+RESET client_min_messages;
+DROP TABLE temp_limit_test_1;
```
Comment (metdos) [inline src/test/regress/sql/multi_complex_expressions.sql]: I think we can skip dropping temp tables for the purpose of clarity.

File: src/test/regress/multi_schedule [hunk 24]
```diff
@@ -29,7 +29,7 @@ test: multi_stage_data
 # ----------
 # Miscellaneous tests to check our query planning behavior
 # ----------
-test: multi_basic_queries multi_complex_expressions multi_verify_no_subquery
+test: multi_basic_queries multi_verify_no_subquery multi_complex_expressions
```
Comment (metdos) [inline src/test/regress/multi_schedule]: Maybe we can keep the original order to not confuse future readers of the changeset.

**Comment by metdos** [comment]
@aamederen, I have a comment about the case where limit is null, and three minor comments (mostly visual). I think we are very close to ship :)

File: src/backend/distributed/planner/multi_logical_optimizer.c (hunk 21 above)
Comment (aamederen) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: Yes, it was a silly mistake 😅  assuming limit = 0 when limit is not present. Changed accordingly.

File: src/backend/distributed/planner/multi_logical_optimizer.c (hunk 22 above)
Comment (aamederen) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: Correct, moved it.

File: src/test/regress/sql/multi_complex_expressions.sql (hunk 23 above)
Comment (aamederen) [inline src/test/regress/sql/multi_complex_expressions.sql]: I generally tend to favor cleaning the mess asap but if you say so, let them go away 👍

File: src/test/regress/multi_schedule (hunk 24 above)
Comment (aamederen) [inline src/test/regress/multi_schedule]: Oh you are correct, after removing the multi_complex_expressions from the parallel group and re-adding it, I forgot to consider their previous order.

**Comment by aamederen** [comment]
There seems to be a problem when LIMIT is not present but there is an OFFSET clause. I am going to work on it.

**Comment by aamederen** [comment]
@metdos I have added a fix to OFFSET without LIMIT issue. Could you please review it?

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 25]
```diff
@@ -4104,38 +4143,61 @@ WorkerLimitCount(MultiExtendedOp *originalOpNode)
...
-		if (orderByNonAggregates)
-		{
-			workerLimitCount = originalOpNode->limitCount;
-		}
-		else if (canApproximate)
-		{
-			Const *workerLimitConst = (Const *) copyObject(originalOpNode->limitCount);
-			int64 workerLimitCountInt64 = (int64) LimitClauseRowFetchCount;
-			workerLimitConst->constvalue = Int64GetDatum(workerLimitCountInt64);
+	/* create the workerLimitNode according to the made decision */
```
Comment (metdos) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: Minor. This does not sound verbally right. Maybe 
..... according to the decision above

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 26]
```diff
@@ -4080,16 +4100,35 @@ EqualOpExpressionLists(List *firstOpExpressionList, List *secondOpExpressionList
...
+ *           has order by agg?          (no pd)
+ *            1/           \0
+ *     can approximate?    (exact pd)
+ *      1/       \0
+ * (approx pd)   (no pd)
+ *
+ * When an offset is present, the offset value is added to limit because for a query
+ * with LIMIT x OFFSET y, (x+y) records should be pulled from the workers.
+ *
+ * If no limit or offset is present or can be pushed down, then WorkerLimitCount
```
Comment (metdos) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: If there is no limit, if though there is offset, we still return null, right?

File: src/test/regress/expected/multi_complex_expressions.out [hunk 27]
```diff
@@ -347,3 +347,115 @@ SELECT count(*) FROM lineitem JOIN orders ON l_orderkey = o_orderkey
 SELECT count(*) FROM lineitem, orders WHERE l_orderkey + 1 = o_orderkey;
 ERROR:  cannot perform local joins that involve expressions
 DETAIL:  local joins can be performed between columns only
+-- Check that we can issue limit/offset queries
+-- OFFSET in Subqueries are not supported
```
Comment (metdos) [inline src/test/regress/expected/multi_complex_expressions.out]: Minor.
Subqueries -> subqueries

File: src/test/regress/sql/multi_complex_expressions.sql [hunk 28]
```diff
@@ -158,3 +158,68 @@ SELECT count(*) FROM lineitem JOIN orders ON l_orderkey = o_orderkey
 -- Check that we make sure local joins are between columns only.
 
 SELECT count(*) FROM lineitem, orders WHERE l_orderkey + 1 = o_orderkey;
+
+-- Check that we can issue limit/offset queries
+
+-- OFFSET in Subqueries are not supported
+-- Error in the planner when subquery pushdown is off
+SELECT * FROM (SELECT o_orderkey FROM orders ORDER BY o_orderkey OFFSET 20) sq;
```
Comment (metdos) [inline src/test/regress/sql/multi_complex_expressions.sql]: Here, new lines between queries would be nice.

**Comment by metdos** [comment]
Changes look good. Squash and 🚢 !
